### ISSUE TITLE
Persistent congestion period: text and pseudo-code don't agree

### ISSUE THREAD

**Issue by janaiyengar** [quicwg/base-drafts#3972]
We resolved #3875 with #3889, but I realized that it has a problem. The text and the pseudo-code don't agree.

The text says:
```
The persistent congestion period SHOULD NOT start until there is at least one RTT sample.
```
Since the `persistent congestion period` starts with a packet that was sent in the past, this statement means that persistent congestion only includes packets that were _sent_ after an RTT sample was taken.

The pseudo code says:
```
   InPersistentCongestion(largest_lost):
     // Persistent congestion cannot be declared on the
     // first RTT sample.
     if (is first RTT sample):
       return false
```
This says that the persistent congestion can only be _established_ on the second (or later) ack that yields an RTT sample. This allows for packets that were sent before the first RTT sample was taken to be included in the persistent congestion period.

I had convinced myself that the pseudo-code was adequate, but I missed the fact that the text was not aligned.

**Comment by janaiyengar** [comment]
I see two options.

Option 1: Implementing what the text says will require a new variable that records the next PN when the first RTT sample is taken.

Option 2: Changing the text to match the pseudo-code will look something like this:
```
Persistent congestion period SHOULD NOT be established until there is at least
one RTT sample.
```

**Comment by janaiyengar** [comment]
An example might help:
1: Pkt 10, 11, 12 sent
2: PTO, Pkt 13, 14 sent
3: PTO, Pkt 15, 16 sent
4: Ack 10 received.
5: Ack 16 received.

At 5, the pseudo-code would declare persistent congestion if `(t3 - t1) > persistent congestion duration`.

The text however wouldn't, since there was no RTT sample during the persistent congestion period, t1 to t3.

**Comment by janaiyengar** [comment]
Option 1 and 2 have a behavior difference with the following implication:

With option 2, a sender might mark persistent congestion with fewer PTOs than if the sender had a more accurate RTT estimate. This is arguably not a significant concern, since:
- the sender will still have sent at least two packets in that duration and marked them lost, since you need two ack-eliciting packets to be lost to declare persistent congestion; and
- the number of packets sent is also sensitive to how the application sends, and a sender cannot rely on that anyway.

Option 2 has the benefit that it is simpler to implement.

On balance, I think Option 2 is better, though it does make this a design change.

**Comment by ianswett** [comment]
Thanks for the example, I was unclear on the issue up until then, though I think a slightly stronger example might be:
 5. Ack 14 received, since that's only a single PTO.

I think 1 aligns more with fully fixing the original issue, since it seems like your example is a case we agreed we would not want to declare as persistent congestion in #3875.  That being said, it makes the pseudocode more complex.

Persistent congestion is a new concept, so I suspect implementations will vary anyway.

**Comment by marten-seemann** [comment]
I agree that making the change to the pseudo-code seems to be the more correct solution.

That being said, I'm not planning to implement persistent congestion detection during the early phases of the handshake (but only after confirmation of the handshake), since I don't see clear benefits of the added complexity of cross-packet-number-space persistent congestion detection.

**Comment by ianswett** [comment]
@marten-seemann I personally think that's fine, though I'll note that TCP is less aggressive than that, which is why I think we should describe this as something that covers the handshake as well.

**Comment by ianswett** [comment]
Given what I've heard, I think this is heading in an editorial direction(fix the pseudocode), though that needs a PR.

**Comment by janaiyengar** [comment]
Yes, I'm not wedded to the design change, and am happy for this to be an editorial one. I'll write up a PR.

**Comment by larseggert** [comment]
Should this become proposal-ready?

**Comment by janaiyengar** [comment]
@larseggert : Since this is a pseudo-code issue, it should be an editorial change (the pseudo code is in the appendix). #4010 will fix this issue, but that is waiting on #3961, which is included in the latest consensus call.
